source/Octopus.Tentacle.Client/Execution/RpcCallExecutor.cs

```diff
@@ -89,28 +90,67 @@ internal RpcCallExecutor(RpcCallRetryHandler rpcCallRetryHandler, ITentacleClien
             Func<CancellationToken, T> action,
             ClientOperationMetricsBuilder clientOperationMetricsBuilder,
             CancellationToken cancellationToken)
+        {
+            return Execute(rpcCall, action, abandonActionOnCancellation: false, clientOperationMetricsBuilder, cancellationToken);
+        }
+
+        public T Execute<T>(
+            RpcCall rpcCall,
+            Func<CancellationToken, T> action,
+            bool abandonActionOnCancellation,
+            ClientOperationMetricsBuilder clientOperationMetricsBuilder,
+            CancellationToken cancellationToken)
         {
             var rpcCallMetricsBuilder = RpcCallMetricsBuilder.StartWithoutRetries(rpcCall);
             var start = DateTimeOffset.UtcNow;
+            var abandonAfter = TimeSpan.FromSeconds(5);
 
-            try
+            T ExecuteAction(Action<Exception> exceptionHandler)
             {
-                var response = action(cancellationToken);
+                try
+                {
+                    var response = action(cancellationToken);
 
-                rpcCallMetricsBuilder.WithAttempt(TimedOperation.Success(start));
-                return response;
+                    rpcCallMetricsBuilder.WithAttempt(TimedOperation.Success(start));
+                    return response;
+                }
+                catch (Exception e)
+                {
+                    rpcCallMetricsBuilder.WithAttempt(TimedOperation.Failure(start, e, cancellationToken));
+                    rpcCallMetricsBuilder.Failure(e, cancellationToken);
+
+                    exceptionHandler(e);
+
+                    throw;
+                }
+                finally
+                {
+                    var rpcCallMetrics = rpcCallMetricsBuilder.Build();
+                    clientOperationMetricsBuilder.WithRpcCall(rpcCallMetrics);
+                    tentacleClientObserver.RpcCallCompleted(rpcCallMetrics);
+                }
             }
-            catch (Exception e)
+
+            if (!abandonActionOnCancellation)
             {
-                rpcCallMetricsBuilder.WithAttempt(TimedOperation.Failure(start, e, cancellationToken));
-                rpcCallMetricsBuilder.Failure(e, cancellationToken);
-                throw;
+                return ExecuteAction(_ =>
```

Comment:
Nit: Make exceptionHandler nullable so we can omit `_ =>{}` which is a bit confusing to read. and then `exceptionHandler?.Invoke(e);`